core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -0,0 +1,719 @@
+            this.sourceExchanges = ImmutableMap.copyOf(requireNonNull(sourceExchanges, "sourceExchanges is null"));
+            this.exchangeSourceHandles = ImmutableListMultimap.copyOf(requireNonNull(exchangeSourceHandles, "exchangeSourceHandles is null"));
+            this.targetPartitionSizeInBytes = requireNonNull(targetPartitionSize, "targetPartitionSize is null").toBytes();
+        }
+
+        @Override
+        public List<TaskDescriptor> getMoreTasks()
+        {
+            NodeRequirements nodeRequirements = new NodeRequirements(Optional.empty(), Optional.empty());
+
+            ImmutableList.Builder<TaskDescriptor> result = ImmutableList.builder();
+            int currentPartitionId = 0;
+
+            ImmutableListMultimap.Builder<PlanNodeId, ExchangeSourceHandle> assignedExchangeSourceHandles = ImmutableListMultimap.builder();
+            long assignedExchangeDataSize = 0;
+
+            for (Map.Entry<PlanFragmentId, ExchangeSourceHandle> entry : exchangeSourceHandles.entries()) {
+                PlanFragmentId sourceFragmentId = entry.getKey();
+                PlanNodeId remoteSourcePlanNodeId = sourceFragmentToRemoteSourceNodeIdMap.get(sourceFragmentId);
+                ExchangeSourceHandle originalExchangeSourceHandle = entry.getValue();
+                Exchange sourceExchange = sourceExchanges.get(sourceFragmentId);
+                ExchangeSourceStatistics exchangeSourceStatistics = sourceExchange.getExchangeSourceStatistics(originalExchangeSourceHandle);
+
+                List<ExchangeSourceHandle> exchangeSourceHandles;
+                if (exchangeSourceStatistics.getSizeInBytes() > targetPartitionSizeInBytes) {
+                    ExchangeSourceSplitter splitter = sourceExchange.split(originalExchangeSourceHandle, targetPartitionSizeInBytes);
+                    ImmutableList.Builder<ExchangeSourceHandle> sourceHandles = ImmutableList.builder();
+                    while (true) {
+                        checkState(splitter.isBlocked().isDone(), "not supported");
+                        Optional<ExchangeSourceHandle> next = splitter.getNext();
+                        if (next.isEmpty()) {
+                            break;
+                        }
+                        sourceHandles.add(next.get());
+                    }
+                    exchangeSourceHandles = sourceHandles.build();
+                }
+                else {
+                    exchangeSourceHandles = ImmutableList.of(originalExchangeSourceHandle);
+                }
+
+                for (ExchangeSourceHandle handle : exchangeSourceHandles) {
+                    ExchangeSourceStatistics statistics = sourceExchange.getExchangeSourceStatistics(handle);
+                    if (assignedExchangeDataSize != 0 && assignedExchangeDataSize + statistics.getSizeInBytes() > targetPartitionSizeInBytes) {
```

Comment:
It feels like we are doing the split twice. Why not rearrange all the source files? `getExchangeSourceStatistics` is not free on the cloud. Listing each object twice is not a good idea.

Reply:
Discussed offline. Going to remove the first size check and always split.